specification/dartLangSpec.tex

```diff
@@ -5193,52 +5195,70 @@ \subsubsection{Correct Member Overrides}
 \begin{itemize}
 \item
   $m$ and $m'$ are both methods, both getters, or both setters.
-  % We cannot have a setter and a method, say, because they cannot have
-  % the same name. However, we must _allow_ $m$ and $m'$ to be two setters.
+\item
+  If $m$ and $m'$ are both getters:
+  The return type of $m$ must be a subtype of the return type of $m'$.
 \item
   If $m$ and $m'$ are both methods or both setters:
   Let $F$ be the function type of $m$
   except that the parameter type is the built-in class \code{Object}
-  for each parameter of $m$ which has the modifier \COVARIANT.
+  for each parameter of $m$ which is covariant-by-declaration
+  (\ref{covariantParameters}).
   Let $F'$ be the function type of $m'$.
   $F$ must then be a subtype of $F'$.
 
   \commentary{%
-  The subtype requirement ensures that argument list shapes
-  that are admissible for an invocation of a method with signature $m'$
-  are also admissible for an invocation of a method with signature $m$.
-  For instance, $m'$ may accept 2 or 3 positional arguments,
-  and $m$ may accept 1, 2, 3, or 4 positional arguments, but not vice versa.
-  This is a built-in property of the function type subtype rules.
-  %
-  Note that a member signature differs from
-  an underlying syntactic declaration $D$ in a class $C$.
-  In particular,
-  a parameter in a member signature has the modifier \COVARIANT{}
-  if and only if the parameter is covariant-by-declaration
-  (\ref{covariantParameters}),
-  and that may be the case due to declarations in a supertype of $C$,
-  so that modifier need not be present in $D$.
-  %
-  There is an additional potential compile-time error associated with
-  a parameter which is covariant-by-declaration
-  (\ref{instanceMethods}).
-  But we cannot cover that here as a property of member overrides,
-  because it is concerned with declarations in all superinterfaces,
-  indirect as well as direct.%
+    The subtype requirement ensures that argument list shapes
+    that are admissible for an invocation of a method with signature $m'$
+    are also admissible for an invocation of a method with signature $m$.
+    For instance, $m'$ may accept 2 or 3 positional arguments,
+    and $m$ may accept 1, 2, 3, or 4 positional arguments, but not vice versa.
+    This is a built-in property of the function type subtype rules.
+  }
+
+  If $p'$ is a formal parameter in $m'$ which is covariant-by-declaration,
+  the corresponding parameter $p$ in $m$ must also be covariant-by-declaration
+  unless the type of $p$ is a top type
+  (\ref{superBoundedTypes}).
+
+  \commentary{%
+    This requirement can actually be violated:
```

Comment:
This phrase is confusing. Does it say that you can write valid programs which violates the rule? Or just that the rule is required because it's possible to write programs where it matters.
I'm guessing the latter, but that's not how I read it.

So, maybe:

>  This requirement matters in the situation where ....

Reply:
I rewrote that commentary again, I think it's considerably more straightforward now! ;-)

It it intended to say that it is possible to write a program where the correct override relation does not hold, because this criterion is not satisfied. If that hadn't been the case then we wouldn't need the criterion at all.

When the correct override relation doesn't hold, it is specified elsewhere that it is an error:

We [specify in section 'Instance Methods'](https://github.com/dart-lang/language/blob/a3f50843c5937135f8542d04aa5a0e239158df7d/specification/dartLangSpec.tex#L2609) that it is an error if a class `C` declares a member, and an accessible member of the same name exists in a superinterface of `C`, and it is not a correct override. So that's the location where we connect 'correct override' and 'compile-time error' for the standard override relation.

[Section 'Classes'](https://github.com/dart-lang/language/blob/a3f50843c5937135f8542d04aa5a0e239158df7d/specification/dartLangSpec.tex#L2522) specifies that an inherited member in a concrete class must be a correct override of any members with the same name in superinterfaces. So that's the location where we connect 'correct override' and 'error' for inherited methods.

Reply:
I think this thread is done, too.